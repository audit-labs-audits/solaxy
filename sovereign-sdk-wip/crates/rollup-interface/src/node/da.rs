//! The da module defines traits used by the full node to interact with the DA layer.

use core::fmt::{Debug, Display};
use std::time::Duration;

use backon::{BackoffBuilder, Retryable};
use schemars::JsonSchema;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use tokio::sync::oneshot;
use tracing::error;

use crate::common::HexHash;
use crate::da::{BlockHeaderTrait, DaSpec, DaVerifier, RelevantBlobs, RelevantProofs, Time};

/// Perform a checked arithmetic, returning None if the result is invalid.
pub trait CheckedMath<Rhs = Self> {
    /// The output type of arithmetic operations
    type Output;

    /// Performs checked multiplication, returning None if the result would overflow.
    fn checked_mul(&self, rhs: Rhs) -> Option<Self::Output>;
    /// Performs checked division, returning None if the caller attempts divide by zero or if the calculation overflows.
    fn checked_div(&self, rhs: Rhs) -> Option<Self::Output>;
    /// Performs checked subtraction, returning None if the result would underflow.
    fn checked_sub(&self, rhs: Rhs) -> Option<Self::Output>;
    /// Performs checked addition, returning None if the result would overflow.
    fn checked_add(&self, rhs: Rhs) -> Option<Self::Output>;
}

macro_rules! impl_checked_math_primitive {
    ($($t:ty),*) => {
        $(impl CheckedMath for $t {
            type Output = Self;

            fn checked_mul(&self, rhs: Self) -> Option<Self::Output> {
                <$t>::checked_mul(*self, rhs)
            }

            fn checked_div(&self, rhs: Self) -> Option<Self::Output> {
                <$t>::checked_div(*self, rhs)
            }

            fn checked_sub(&self, rhs: Self) -> Option<Self::Output> {
                <$t>::checked_sub(*self, rhs)
            }

            fn checked_add(&self, rhs: Self) -> Option<Self::Output> {
                <$t>::checked_add(*self, rhs)
            }
        })*
    };
}

impl_checked_math_primitive!(u8, u16, u32, u64, u128, usize);
impl_checked_math_primitive!(i8, i16, i32, i64, i128, isize);

/// The [`MaybeRetryable`] enum can be returned from a fallible function to
/// determine whether it can re-attempted or not.
#[derive(Debug, thiserror::Error)]
pub enum MaybeRetryable<E> {
    /// This error is a permanent one and thus the function that
    /// raised it should not be retried.
    #[error("{0}")]
    Permanent(E),
    /// This error is transient and thus the function that raised
    /// ought to be retried.
    #[error("{0}")]
    Transient(E),
}

impl<E: std::fmt::Display> MaybeRetryable<E> {
    fn is_retryable(&self) -> bool {
        matches!(self, Self::Transient(_))
    }

    fn into_err(self) -> E {
        match self {
            Self::Permanent(e) | Self::Transient(e) => e,
        }
    }
}

/// The default error for `MaybeRetryable` is assumed to be permanent
/// rather than something that can be retried.
impl<E> From<E> for MaybeRetryable<E> {
    fn from(e: E) -> MaybeRetryable<E> {
        Self::Permanent(e)
    }
}

/// Output of submit blob operation.
#[derive(Debug, Clone, Serialize, Deserialize, derive_more::Display)]
#[display(
    "SubmitBlobReceipt {{ blob_hash: {}, da_transaction_id: {:?} }}",
    blob_hash,
    da_transaction_id
)]
pub struct SubmitBlobReceipt<T: Debug + Clone> {
    /// Computed blob hash, so it can be identified by fetcher of the blobs.
    pub blob_hash: HexHash,
    /// Identifier of the transaction on the DA layer.
    pub da_transaction_id: T,
}

/// A DaService is the local side of an RPC connection talking to a node of the DA layer
/// It is *not* part of the logic that is zk-proven.
///
/// The DaService has two responsibilities - fetching data from the DA layer, transforming the
/// data into a representation that can be efficiently verified in circuit.
#[async_trait::async_trait]
pub trait DaService: Clone + Send + Sync + 'static {
    /// A handle to the types used by the DA layer.
    type Spec: DaSpec;

    /// [`serde`]-compatible configuration data for this [`DaService`]. Parsed
    /// from TOML.
    type Config: JsonSchema + PartialEq + Send + Sync + 'static;

    /// The verifier for this DA layer.
    type Verifier: DaVerifier<Spec = Self::Spec> + Clone;

    /// A DA layer block, possibly excluding some irrelevant information.
    type FilteredBlock: SlotData<BlockHeader = <Self::Spec as DaSpec>::BlockHeader>;

    /// The error type for fallible methods.
    type Error: Debug + Send + Sync + Display;

    /// Subsequent calls of [`DaService::send_transaction`] guarantee that the
    /// transactions are published and land in the DA layer in the same order as
    /// the method calls.
    const GUARANTEES_TRANSACTION_ORDERING: bool = false;

    /// Fetch the block at the given height, waiting for one to be mined if necessary.
    ///
    /// The returned block may not be final, and can be reverted without a consensus violation.
    /// Calls to this method for the same height are allowed to return different results.
    /// Should always returns the block at that height on the best fork.
    async fn get_block_at(&self, height: u64) -> Result<Self::FilteredBlock, Self::Error>;

    /// Similar to [`DaService::get_block_at`], but only returns the block header and not the whole block.
    /// All constraints and limitations of [`Self::get_block_at`] apply
    async fn get_block_header_at(
        &self,
        height: u64,
    ) -> Result<<Self::Spec as DaSpec>::BlockHeader, Self::Error> {
        let block = self.get_block_at(height).await?;
        Ok(block.header().clone())
    }

    /// How long the node should wait after a block is produced before
    /// submitting a transaction.
    ///
    /// The returned value must be low for the node to be reasonably confident
    /// that the transaction will be included in the next block.
    ///
    /// The [`DaService`] is allowed to adjust this value based on network
    /// conditions. Set to [`Duration::ZERO`] by default.
    fn safe_lead_time(&self) -> Duration {
        Duration::ZERO
    }

    /// Fetch the [`DaSpec::BlockHeader`] of the last finalized block.
    /// If there's no finalized block yet, it should return an error.
    async fn get_last_finalized_block_header(
        &self,
    ) -> Result<<Self::Spec as DaSpec>::BlockHeader, Self::Error>;

    /// Fetch the height of the last finalized block.
    ///
    /// ## Why not just use [`DaService::get_last_finalized_block_header`]?
    ///
    /// Some [`DaService`] implementations may have a way of querying for the
    /// latest finalized block number that's faster than fetching the whole
    /// header. This method allows for such [`DaService`]-specific
    /// optimizations.
    async fn get_last_finalized_block_number(&self) -> Result<u64, Self::Error> {
        Ok(self.get_last_finalized_block_header().await?.height())
    }

    /// Fetch the head block of the most popular fork.
    ///
    /// More like utility method, to provide better user experience
    async fn get_head_block_header(
        &self,
    ) -> Result<<Self::Spec as DaSpec>::BlockHeader, Self::Error>;

    /// Extract the relevant transactions from a block. For example, this method might return
    /// all of the blob transactions from a set rollup namespaces on Celestia.
    fn extract_relevant_blobs(
        &self,
        block: &Self::FilteredBlock,
    ) -> RelevantBlobs<<Self::Spec as DaSpec>::BlobTransaction>;

    /// Generate a proof that the relevant blob transactions have been extracted correctly from the DA layer
    /// block.
    async fn get_extraction_proof(
        &self,
        block: &Self::FilteredBlock,
        blobs: &RelevantBlobs<<Self::Spec as DaSpec>::BlobTransaction>,
    ) -> RelevantProofs<
        <Self::Spec as DaSpec>::InclusionMultiProof,
        <Self::Spec as DaSpec>::CompletenessProof,
    >;

    /// Extract the relevant transactions from a block, along with a proof that the extraction has been done correctly.
    /// For example, this method might return all of the blob transactions in rollup's namespace on Celestia,
    /// together with a range proof against the root of the namespaced-merkle-tree, demonstrating that the entire
    /// rollup namespace has been covered.
    #[allow(clippy::type_complexity)]
    async fn extract_relevant_blobs_with_proof(
        &self,
        block: &Self::FilteredBlock,
    ) -> (
        RelevantBlobs<<Self::Spec as DaSpec>::BlobTransaction>,
        RelevantProofs<
            <Self::Spec as DaSpec>::InclusionMultiProof,
            <Self::Spec as DaSpec>::CompletenessProof,
        >,
    ) {
        let relevant_blobs = self.extract_relevant_blobs(block);
        let relevant_proofs = self.get_extraction_proof(block, &relevant_blobs).await;
        (relevant_blobs, relevant_proofs)
    }

    /// Send a transaction directly to the DA layer.
    /// This method is infallible: the SubmitBlobReceipt is returned via the `oneshot::Receiver` after the blob is posted to the DA.
    async fn send_transaction(
        &self,
        blob: &[u8],
    ) -> oneshot::Receiver<
        Result<SubmitBlobReceipt<<Self::Spec as DaSpec>::TransactionId>, Self::Error>,
    >;

    /// Sends a proof to the DA layer.
    /// This method is infallible: the SubmitBlobReceipt is returned via the `oneshot::Receiver` after the blob is posted to the DA.
    async fn send_proof(
        &self,
        aggregated_proof_data: &[u8],
    ) -> oneshot::Receiver<
        Result<SubmitBlobReceipt<<Self::Spec as DaSpec>::TransactionId>, Self::Error>,
    >;

    /// Fetches all proofs at a specified block height.
    async fn get_proofs_at(&self, height: u64) -> Result<Vec<Vec<u8>>, Self::Error>;

    /// Returns a [`tokio::task::JoinHandle`] to the DA service background task,
    /// if it exists.
    async fn take_background_join_handle(&self) -> Option<tokio::task::JoinHandle<()>> {
        None
    }
}

/// Retry the given async function with the given backoff policy.
pub async fn run_maybe_retryable_async_fn_with_retries<F, Fut, T, E>(
    backoff_policy: &impl BackoffBuilder,
    fxn: F,
    da_method_name: &str,
) -> Result<T, E>
where
    F: Fn() -> Fut,
    Fut: std::future::Future<Output = Result<T, MaybeRetryable<E>>>,
    E: std::fmt::Display,
{
    fxn.retry(backoff_policy)
        .notify(|err: &MaybeRetryable<E>, dur: Duration| {
            tracing::warn!(
                method_name = da_method_name, error = %err, duration = ?dur,
                "Error in DA Service, will retry in specified duration."
            );
        })
        .when(MaybeRetryable::is_retryable)
        .await
        .map_err(MaybeRetryable::into_err)
}

/// `SlotData` is the subset of a DA layer block which is stored in the rollup's database.
/// At the very least, the rollup needs access to the hashes and headers of all DA layer blocks,
/// but rollup may choose to store partial (or full) block data as well.
pub trait SlotData:
    Serialize + DeserializeOwned + PartialEq + core::fmt::Debug + Clone + Send + Sync
{
    /// The header type for a DA layer block as viewed by the rollup. This need not be identical
    /// to the underlying rollup's header type, but it must be sufficient to reconstruct the block hash.
    ///
    /// For example, most fields of a Tendermint-based DA chain like Celestia are irrelevant to the rollup.
    /// For these fields, we only ever store their *serialized* representation in memory or on disk. Only a few special
    /// fields like `data_root` are stored in decoded form in the `CelestiaHeader` struct.
    type BlockHeader: BlockHeaderTrait;

    /// The canonical hash of the DA layer block.
    fn hash(&self) -> [u8; 32];
    /// The header of the DA layer block.
    fn header(&self) -> &Self::BlockHeader;
    /// The timestamp of the DA layer block.
    fn timestamp(&self) -> Time;
}

#[cfg(test)]
mod tests {
    use backon::ExponentialBuilder;

    use super::*;

    #[tokio::test(flavor = "multi_thread")]
    async fn should_run_async_fn_with_retries() {
        let error = "some error".to_string();
        let retry_counter = tokio::sync::Mutex::new(0);
        let max_retries = 3;
        let backoff_policy = ExponentialBuilder::default().with_max_times(max_retries);

        let r = run_maybe_retryable_async_fn_with_retries(
            &backoff_policy,
            || async {
                let mut count = retry_counter.lock().await;
                *count += 1;
                Result::<(), MaybeRetryable<String>>::Err(MaybeRetryable::Transient(error.clone()))
            },
            "test_function",
        )
        .await;

        assert_eq!(r, Err(error));
        assert_eq!(*retry_counter.lock().await, max_retries + 1); // NOTE: Because first attempt is not a retry.
    }
}
