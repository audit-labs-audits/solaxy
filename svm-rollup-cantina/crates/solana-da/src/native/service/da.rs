use anyhow::anyhow;
use futures::FutureExt;
use jsonrpsee::core::async_trait;
use solana_sdk::commitment_config::CommitmentConfig;
use sov_rollup_interface::{
    da::{DaSpec, RelevantBlobs, RelevantProofs},
    node::da::{DaService, MaybeRetryable, SubmitBlobReceipt},
};
use tokio::sync::oneshot;
use tracing::info;

use crate::{
    native::service::{client_error_to_maybe_retryable, SolanaService},
    SolanaBlock, SolanaBlockHeader, SolanaDASpec, SolanaServiceConfig, SolanaVerifier,
};

#[async_trait]
impl DaService for SolanaService {
    type Spec = SolanaDASpec;
    type Config = SolanaServiceConfig;
    type Verifier = SolanaVerifier;
    type FilteredBlock = SolanaBlock;
    type Error = anyhow::Error;

    /// Fetch the block at the given height, waiting for one to be mined if necessary.
    /// The returned block may not be final, and can be reverted without a consensus violation.
    /// Call it for the same height are allowed to return different results.
    /// Should always returns the block at that height on the best fork.
    async fn get_block_at(&self, height: u64) -> Result<Self::FilteredBlock, Self::Error> {
        let block = self.rpc_get_block(height).await?;

        let batch = self
            .get_indexer_data(height, "batch", &self.params.blober_batch)
            .boxed();
        let proof = self
            .get_indexer_data(height, "proof", &self.params.blober_proofs)
            .boxed();

        let (batch_data, proof_data) = futures::future::try_join(batch, proof).await?;

        Ok(SolanaBlock::new(height, block, batch_data, proof_data))
    }

    /// Subscribe to finalized headers as they are finalized.
    /// Expect only to receive headers which were finalized after subscription
    /// Optimized version of `get_last_finalized_block_header`.
    async fn get_last_finalized_block_header(
        &self,
    ) -> Result<<Self::Spec as DaSpec>::BlockHeader, Self::Error> {
        info!("get_last_finalized_block_header START");
        let slot = self
            .rpc()
            .get_slot_with_commitment(CommitmentConfig::finalized())
            .await
            .map_err(client_error_to_maybe_retryable(
                "Problem when retrieving last finalized slot from Solana RPC",
            ))?;
        let block = self
            .rpc()
            .get_block_with_config(
                slot,
                solana_client::rpc_config::RpcBlockConfig {
                    encoding: None,
                    transaction_details: Some(solana_transaction_status::TransactionDetails::Full),
                    rewards: Some(false),
                    commitment: Some(CommitmentConfig::finalized()),
                    max_supported_transaction_version: None,
                },
            )
            .await
            .map_err(client_error_to_maybe_retryable(
                "Problem when retrieving last finalized block from Solana RPC",
            ))?
            .into();

        info!("get_last_finalized_block_header DONE");
        Ok(SolanaBlockHeader::new(slot, &block))
    }

    /// Fetch the head block of the most popular fork.
    ///
    /// More like utility method, to provide better user experience
    async fn get_head_block_header(
        &self,
    ) -> Result<<Self::Spec as DaSpec>::BlockHeader, Self::Error> {
        info!("get_head_block_header START");
        let slot = self
            .rpc()
            .get_slot_with_commitment(CommitmentConfig::confirmed())
            .await
            .map_err(client_error_to_maybe_retryable(
                "Problem when retrieving last confirmed slot from Solana RPC",
            ))?;
        let block = self
            .rpc()
            .get_block_with_config(
                slot,
                solana_client::rpc_config::RpcBlockConfig {
                    encoding: None,
                    transaction_details: Some(solana_transaction_status::TransactionDetails::Full),
                    rewards: Some(false),
                    commitment: Some(CommitmentConfig::confirmed()),
                    max_supported_transaction_version: None,
                },
            )
            .await
            .map_err(client_error_to_maybe_retryable(
                "Problem when retrieving last confirmed block from Solana RPC",
            ))?
            .into();

        info!("get_head_block_header DONE");
        Ok(SolanaBlockHeader::new(slot, &block))
    }

    /// Extract the relevant transactions from a block.
    fn extract_relevant_blobs(
        &self,
        block: &Self::FilteredBlock,
    ) -> RelevantBlobs<<Self::Spec as DaSpec>::BlobTransaction> {
        info!("extract_relevant_blobs START");
        let relevant_blobs = block.into();
        info!("extract_relevant_blobs DONE");
        relevant_blobs
    }

    /// Generate a proof that the relevant blob transactions have been extracted correctly from the DA layer
    /// block.
    async fn get_extraction_proof(
        &self,
        block: &Self::FilteredBlock,
        _blobs: &RelevantBlobs<<Self::Spec as DaSpec>::BlobTransaction>,
    ) -> RelevantProofs<
        <Self::Spec as DaSpec>::InclusionMultiProof,
        <Self::Spec as DaSpec>::CompletenessProof,
    > {
        info!("get_extraction_proof START");
        let extraction_proof = block.into();
        info!("get_extraction_proof DONE");
        extraction_proof
    }

    /// Send a transaction directly to the DA layer.
    /// blob is the serialized and signed transaction.
    /// Returns transaction id if it was successfully sent.
    async fn send_transaction(
        &self,
        blob: &[u8],
    ) -> tokio::sync::oneshot::Receiver<
        Result<SubmitBlobReceipt<<Self::Spec as DaSpec>::TransactionId>, Self::Error>,
    > {
        let (tx, rx) = oneshot::channel();

        let res = self
            .send_transaction(blob)
            .await
            .map_err(|e| anyhow::anyhow!(e));

        tx.send(res).unwrap();
        rx
    }

    /// Sends a proof to the DA layer.
    async fn send_proof(
        &self,
        aggregated_proof_data: &[u8],
    ) -> tokio::sync::oneshot::Receiver<
        Result<SubmitBlobReceipt<<Self::Spec as DaSpec>::TransactionId>, Self::Error>,
    > {
        let (tx, rx) = oneshot::channel();

        let res = self
            .send_proof(aggregated_proof_data)
            .await
            .map_err(|e| anyhow::anyhow!(e));

        tx.send(res).unwrap();
        rx
    }

    /// Fetches all proofs at a specified block height.
    async fn get_proofs_at(&self, height: u64) -> Result<Vec<Vec<u8>>, Self::Error> {
        info!("get_proofs_at {height} START");
        let blobs = self
            .blober()
            .get_blobs(height, self.params.blober_proofs)
            .await
            .map_err(|err| {
                MaybeRetryable::Permanent(anyhow!("Error when retrieving proof blobs: {err:?}"))
            })?;
        info!("get_proofs_at {height} DONE");
        Ok(blobs)
    }
}
