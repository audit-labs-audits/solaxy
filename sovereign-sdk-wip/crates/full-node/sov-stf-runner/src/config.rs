use std::num::NonZero;
use std::path::{Path, PathBuf};

use schemars::JsonSchema;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use sov_rollup_interface::node::da::DaService;
use sov_sequencer::{SequencerConfig, SequencerKindConfig};

pub const DEFAULT_CONCURRENT_SYNC_TASKS: u8 = 5;
pub use sov_metrics::{MonitoringConfig, TelegrafSocketConfig};

/// Configuration for StateTransitionRunner.
#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct RunnerConfig {
    /// DA start height.
    pub genesis_height: u64,
    /// Polling interval for the DA service to check the sync status (in milliseconds).
    pub da_polling_interval_ms: u64,
    /// HTTP Server configuration: On this socket REST API and RPC endpoints are going to listen.
    pub http_config: HttpServerConfig,
    /// How many concurrent tasks to get block from DA service
    pub concurrent_sync_tasks: Option<u8>,
}

impl RunnerConfig {
    pub(crate) fn get_concurrent_sync_tasks(&self) -> u8 {
        self.concurrent_sync_tasks
            .unwrap_or(DEFAULT_CONCURRENT_SYNC_TASKS)
    }
}

/// Configuration for HTTP server(s) exposed by the node.
#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct HttpServerConfig {
    /// Server host.
    pub bind_host: String,
    /// Server port.
    pub bind_port: u16,
    /// The fully qualified public name of the server, in case the rollup node is running behind a proxy.
    /// For instance:
    /// ```toml
    /// public_address = "https://rollup.example.com"
    /// ```
    pub public_address: Option<String>,
    /// Enable or disable CORS policy headers. Enabled by default.
    #[serde(default)]
    pub cors: CorsConfiguration,
}

/// See [`HttpServerConfig::cors`].
///
/// # Default
///
/// Cross-origin resource sharing (CORS) makes local development easier, so it's enabled by default
///
/// In production, one may want to disable it and let your reverse proxy
/// handle CORS instead.
/// Security note:
/// Allowing CORS in production allows making requests to rollup node any website on the Internet.
#[derive(Debug, Clone, Copy, Default, Deserialize, Serialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum CorsConfiguration {
    /// Enables CORS with a permissive policy, intended for local development.
    #[default]
    Permissive,
    /// Disables CORS.
    Restrictive,
}

impl HttpServerConfig {
    /// Creates an [`HttpServerConfig`] that requests the operating system to bind to any available port.
    /// Useful for testing as it prevents multiple threads from binding to the same port.
    pub fn localhost_on_free_port() -> Self {
        Self::localhost_on_port(0)
    }

    /// Creates an [`HttpServerConfig`] that listens on the provided port using sensible defaults
    /// for local testing.
    pub fn localhost_on_port(port: u16) -> Self {
        HttpServerConfig {
            bind_host: "127.0.0.1".to_string(),
            bind_port: port,
            public_address: None,
            cors: CorsConfiguration::Permissive,
        }
    }

    /// Creates an [`HttpServerConfig`] that listens on provided host and port,
    /// using sensible default for local testing
    pub fn on_host_port(host: impl Into<String>, port: u16) -> Self {
        HttpServerConfig {
            bind_host: host.into(),
            bind_port: port,
            public_address: None,
            cors: CorsConfiguration::Permissive,
        }
    }
}

/// Simple storage configuration
#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct StorageConfig {
    /// Path that can be utilized by concrete implementation
    pub path: PathBuf,
}

/// Prover service configuration.
#[derive(Debug, Clone, Deserialize, Serialize, Copy, JsonSchema)]
pub struct ProofManagerConfig<Address> {
    /// The "distance" measured in the number of blocks between two consecutive aggregated proofs.
    pub aggregated_proof_block_jump: NonZero<usize>,
    /// The prover receives rewards to this address.
    pub prover_address: Address,
    /// A number of state transition info entries are allowed to be stored in the database.
    /// When the number is exceeded, older entries are removed.
    pub max_number_of_transitions_in_db: NonZero<u64>,
    /// A number of state transition info entries are allowed to be kept in memory.
    /// If the number is exceeded, rollup execution will be blocked until provers cathes up.
    pub max_number_of_transitions_in_memory: NonZero<u64>,
}

/// Rollup Configuration
#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
#[schemars(bound = "Address: JsonSchema, Da: DaService", rename = "RollupConfig")]
pub struct RollupConfig<Address, Da: DaService> {
    /// Currently rollup config runner only supports storage path parameter
    pub storage: StorageConfig,
    /// Runner own configuration.
    pub runner: RunnerConfig,
    /// Data Availability service configuration.
    pub da: Da::Config,
    /// Proof manager configuration.
    pub proof_manager: ProofManagerConfig<Address>,
    /// Sequencer (and batch builder) configuration.
    pub sequencer: SequencerConfig<Da::Spec, Address, SequencerKindConfig>,
    /// Monitoring configuration.
    pub monitoring: MonitoringConfig,
}

/// Reads toml file as a specific type.
pub fn from_toml_path<P: AsRef<Path>, R: DeserializeOwned>(path: P) -> anyhow::Result<R> {
    let contents = std::fs::read_to_string(&path)?;

    tracing::info!(
        path = path.as_ref().to_string_lossy().to_string(),
        size_in_bytes = contents.len(),
        line_count = contents.lines().count(),
        "Parsing rollup configuration file"
    );

    Ok(toml::from_str(&contents)?)
}

#[cfg(test)]
mod tests {
    use sov_mock_da::MockDaService;
    use sov_modules_api::Address;

    use crate::RollupConfig;

    #[test]
    fn test_correct_config() {
        let config_s = r#"
            [da]
            connection_string = "sqlite:///tmp/mockda.sqlite?mode=rwc"
            sender_address = "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f"
            [da.block_producing.periodic]
            block_time_ms = 1_000
            [storage]
            path = "/tmp"
            [runner]
            genesis_height = 31337
            da_polling_interval_ms = 10000
            concurrent_sync_tasks = 18
            [runner.http_config]
            bind_host = "127.0.0.1"
            bind_port = 12346
            public_address = "https://rollup.sovereign.xyz"
            cors = "restrictive"
            [monitoring]
            telegraf_address = "udp://192.168.4.5:8543"
            max_datagram_size = 1024
            max_pending_metrics = 2560
            [proof_manager]
            aggregated_proof_block_jump = 22
            prover_address = "sov1lzkjgdaz08su3yevqu6ceywufl35se9f33kztu5cu2spja5hyyf"
            max_number_of_transitions_in_db = 1025
            max_number_of_transitions_in_memory = 768
            [sequencer]
            max_batch_size_bytes = 1048576
            max_concurrent_blobs = 16
            max_allowed_node_distance_behind = 5
            da_address = "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f"
            rollup_address = "sov1lzkjgdaz08su3yevqu6ceywufl35se9f33kztu5cu2spja5hyyf"
            [sequencer.standard]
        "#;

        let config = toml::from_str::<RollupConfig<Address, MockDaService>>(config_s).unwrap();
        insta::assert_json_snapshot!(config);
    }
}
