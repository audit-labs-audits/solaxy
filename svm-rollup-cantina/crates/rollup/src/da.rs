#[derive(clap::ValueEnum, Clone, Debug, Copy)]
pub enum SupportedDaLayer {
    Celestia,
    Mock,
    Solana,
}

pub mod celestia {
    use sov_celestia_adapter::{
        types::Namespace,
        verifier::{CelestiaVerifier, RollupParams},
    };
    pub use sov_celestia_adapter::{
        verifier::CelestiaSpec as DaSpec, CelestiaService as DaService,
    };
    use sov_modules_api::{prelude::tokio::sync::watch::Receiver, Spec};
    use sov_rollup_interface::da::DaVerifier;
    use sov_stf_runner::RollupConfig;

    /// The rollup stores its data in the namespace "sov-test-b" on Celestia.
    /// You can change this constant to point your rollup at a different namespace.
    const ROLLUP_BATCH_NAMESPACE: Namespace = Namespace::const_v0(*b"sov-test-b");

    /// The rollup stores the zk proofs in the namespace "sov-test-p" on Celestia.
    const ROLLUP_PROOF_NAMESPACE: Namespace = Namespace::const_v0(*b"sov-test-p");

    pub fn new_verifier() -> CelestiaVerifier {
        CelestiaVerifier::new(RollupParams {
            rollup_batch_namespace: ROLLUP_BATCH_NAMESPACE,
            rollup_proof_namespace: ROLLUP_PROOF_NAMESPACE,
        })
    }

    pub async fn new_da_service<S: Spec>(
        rollup_config: &RollupConfig<S::Address, DaService>,
        _shutdown_receiver: Receiver<()>,
    ) -> DaService {
        DaService::new(
            rollup_config.da.clone(),
            RollupParams {
                rollup_batch_namespace: ROLLUP_BATCH_NAMESPACE,
                rollup_proof_namespace: ROLLUP_PROOF_NAMESPACE,
            },
        )
        .await
    }
}

pub mod mock {
    use sov_mock_da::MockDaVerifier;
    pub use sov_mock_da::{
        storable::service::StorableMockDaService as DaService, MockDaSpec as DaSpec,
    };
    use sov_modules_api::{prelude::tokio::sync::watch::Receiver, Spec};
    use sov_stf_runner::RollupConfig;

    pub fn new_verifier() -> MockDaVerifier {
        MockDaVerifier::default()
    }

    pub async fn new_da_service<S: Spec>(
        rollup_config: &RollupConfig<S::Address, DaService>,
        shutdown_receiver: Receiver<()>,
    ) -> DaService {
        DaService::from_config(rollup_config.da.clone(), shutdown_receiver).await
    }
}

pub mod solana {
    use std::sync::Arc;

    use nitro_da_client::BloberClient;
    use solana_cli_config::Config;
    pub use solana_da::{
        SolanaChainParams, SolanaDASpec as DaSpec, SolanaService as DaService, SolanaVerifier,
    };
    use solana_program::pubkey::Pubkey;
    use solana_sdk::{
        signature::{read_keypair_file, Keypair},
        signer::Signer,
    };
    use sov_modules_api::{prelude::tokio::sync::watch::Receiver, Spec};
    use sov_rollup_interface::da::DaVerifier;
    use sov_stf_runner::RollupConfig;

    pub fn new_verifier(program_id: Pubkey, payer: Pubkey) -> SolanaVerifier {
        SolanaVerifier::new(SolanaChainParams::new(program_id, payer))
    }

    struct SolanaDaConfig {
        cli: Config,
        payer: Arc<Keypair>,
    }

    fn read_config() -> SolanaDaConfig {
        let default_file = solana_cli_config::CONFIG_FILE
            .as_ref()
            .expect("unable to get solana config file path")
            .to_owned();

        let solana_config_file = std::env::var("SOLANA_CONFIG_FILE").unwrap_or(default_file);

        let mut cli = Config::load(&solana_config_file).expect("unable to load solana config file");

        if cli.websocket_url.is_empty() {
            cli.websocket_url = Config::compute_websocket_url(&cli.json_rpc_url);
        }

        let payer =
            Arc::new(read_keypair_file(&cli.keypair_path).expect("unable to read keypair file"));

        SolanaDaConfig { cli, payer }
    }

    pub async fn new_da_service<S: Spec>(
        rollup_config: &RollupConfig<S::Address, DaService>,
        _shutdown_receiver: Receiver<()>,
    ) -> DaService {
        let SolanaDaConfig { cli, payer } = read_config();
        tracing::warn!("cli: {cli:?}");

        let program_id = rollup_config.da.program_id.into();

        let blober_client = BloberClient::builder()
            .payer(payer.clone())
            .program_id(program_id)
            .indexer_from_url(&rollup_config.da.indexer_url)
            .await
            .expect("to build indexer client")
            .build_with_config(cli)
            .await
            .expect("to build default rpc and batch clients");

        let params = SolanaChainParams::new(program_id, payer.pubkey());

        tracing::warn!("params: {params:?}");

        let config = rollup_config.da.clone();

        DaService::builder()
            .solana_rpc_client(blober_client.rpc_client())
            .blober_client(blober_client)
            .params(params)
            .config(config)
            .build()
    }
}
