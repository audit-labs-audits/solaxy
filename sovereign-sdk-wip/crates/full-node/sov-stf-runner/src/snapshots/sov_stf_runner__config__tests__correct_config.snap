---
source: crates/full-node/sov-stf-runner/src/config.rs
expression: config
---
{
  "storage": {
    "path": "/tmp"
  },
  "runner": {
    "genesis_height": 31337,
    "da_polling_interval_ms": 10000,
    "http_config": {
      "bind_host": "127.0.0.1",
      "bind_port": 12346,
      "public_address": "https://rollup.sovereign.xyz",
      "cors": "restrictive"
    },
    "concurrent_sync_tasks": 18
  },
  "da": {
    "connection_string": "sqlite:///tmp/mockda.sqlite?mode=rwc",
    "sender_address": "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f",
    "finalization_blocks": 0,
    "block_producing": {
      "periodic": {
        "block_time_ms": 1000
      }
    },
    "randomization": null
  },
  "proof_manager": {
    "aggregated_proof_block_jump": 22,
    "prover_address": "sov1lzkjgdaz08su3yevqu6ceywufl35se9f33kztu5cu2spja5hyyf",
    "max_number_of_transitions_in_db": 1025,
    "max_number_of_transitions_in_memory": 768
  },
  "sequencer": {
    "automatic_batch_production": true,
    "max_allowed_node_distance_behind": 5,
    "dropped_tx_ttl_secs": 60,
    "da_address": "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f",
    "rollup_address": "sov1lzkjgdaz08su3yevqu6ceywufl35se9f33kztu5cu2spja5hyyf",
    "admin_addresses": [],
    "standard": {
      "mempool_max_txs_count": null,
      "max_batch_size_bytes": null
    },
    "max_batch_size_bytes": 1048576,
    "max_concurrent_blobs": 16
  },
  "monitoring": {
    "telegraf_address": "udp://192.168.4.5:8543",
    "max_datagram_size": 1024,
    "max_pending_metrics": 2560
  }
}
